### PR TITLE
feat: Updated and tuned default sqlite config

### PR DESCRIPTION
**Pull request by fedefrancescon**
- Updated the default config with newer defines and removed old, no longer existing, ones.
- Fixed minor compiler warnings
- Polished CMakeLists.txt and removed redundant (apparently ununsed) component.mk
- Added IO Stats to help investigate and debug problems
- Updated SQLite to v3.47.0
- Minor code cleanup

### CODE REVIEW

**Comment by fedefrancescon** [comment]
@fedepell I've assigned it to myself as, before merging, I'd like to complete some tests

File: config_ext.h [hunk 1]
```diff
@@ -78,12 +74,26 @@
...
 /* #define SQLITE_OMIT_FLOATING_POINT        1 */
-#define SQLITE_COUNTOFVIEW_OPTIMIZATION      0
+
+#define SQLITE_OMIT_JSON                     1
+#define SQLITE_OMIT_INTROSPECTION_PRAGMAS    1
+#define SQLITE_OMIT_COMPLETE                 1
+#define SQLITE_OMIT_DESERIALIZE              1
+#define SQLITE_DISABLE_PAGECACHE_OVERFLOW_STATS 1
+#define SQLITE_ENABLE_NULL_TRIM              1
+#define SQLITE_ENABLE_PREUPDATE_HOOK         1 // This may be interesting to implement the update logic
```
Comment (fedepell) [inline config_ext.h]: I was reading about this too but I'm afraid not, as the hook cannot do anything on the database (nor change what has been asked to do). Here is an example/question which may be similar to your idea that should not be feasible with the hook indeed: https://stackoverflow.com/questions/37545766/sqlite-pre-update-hook-re-entrant

File: esp32.c [hunk 2]
```diff
@@ -14,9 +14,10 @@
 #include <time.h>
 #include <unistd.h>
 #include <sqlite3.h>
-#include <esp_spi_flash.h>
+#include <spi_flash_mmap.h>
 #include <esp_system.h>
 #include <esp_random.h>
+#include <esp_timer.h>
```
Comment (fedepell) [inline esp32.c]: I think this is just used for IO_STATS so maybe better:

```suggestion
#ifdef DEBUG_IO_STATS
#include <esp_timer.h>
#endif
```

Of course moving the `#define DEBUG_IO_STATS` with comment on the top?

File: esp32.c [hunk 3]
```diff
@@ -99,59 +85,104 @@ typedef struct esp32_file {
...
-	esp32_Lock,
-	esp32_Unlock,
-	esp32_CheckReservedLock,
-	esp32_FileControl,
-	esp32_SectorSize,
-	esp32_DeviceCharacteristics
+	.iVersion                 = 1,
+	.xClose                   = esp32_Close,
+	#ifdef DEBUG_IO_STATS
+		.xRead                  = esp32_Read,
```
Comment (fedepell) [inline esp32.c]:
```suggestion
		.xRead                  = esp32_Read_Stats,
```

File: esp32.c [hunk 4]
```diff
@@ -644,6 +686,68 @@ int esp32_CurrentTime( sqlite3_vfs * vfs, double * result )
...
+		}
+
+		int64_t start = esp_timer_get_time();
+		int rc = esp32_Read(id, buffer, amount, offset);
+		file->stats.read_elapsed_us += esp_timer_get_time() - start;
+
+		if (rc == SQLITE_OK) {
+			file->stats.read_failed++;
+		} else {
+			file->stats.read_bytes += amount;
```
Comment (fedepell) [inline esp32.c]: Is this the other way around?

```suggestion
		if (rc == SQLITE_OK) {
			file->stats.read_bytes += amount;
		} else {
			file->stats.read_failed++;
```

Comment (fedefrancescon) [reply esp32.c]: Done

File: esp32.c (hunk 2 above)
Comment (fedefrancescon) [reply esp32.c]: Done

File: config_ext.h (hunk 1 above)
Comment (fedefrancescon) [reply config_ext.h]: what a pity!

Removed

**Comment by fedefrancescon** [comment]
Drafted, as you see I've not yet debugged the read part :wink:

**Comment by fedepell** [comment]
Seems like I cannot put an inline comment as it wasn't modified :)

Albeit some tests I did seemed not to give improvements (but were a bit more generic), I'd suggest to put the page size (so `SQLITE_DEFAULT_PAGE_SIZE`) to 4096 (which BTW would be the default if not overridden) given that the SPI flash seems to be also paged at 4096 (see here: https://github.com/espressif/esp-idf/blob/master/components/spi_flash/include/spi_flash_mmap.h#L29)

Maybe with your test and IOStats we can see if there is any difference (ie. smaller number of writes?).

**Comment by fedefrancescon** [comment]
At the end I left the in-memory code, as this seems to have some relevant performance gains.

Still no luck with WAL, but for not I would leave it out.

File: CMakeLists.txt [hunk 5]
```diff
@@ -4,6 +4,31 @@ idf_component_register(SRCS "sqlite3.c" "esp32.c"
...
+    # =============================
+    # @TODO  Double check these
+    # eventually remove or restore
+    # before merging!!
+    # =============================
+
+    # -fno-stack-protector
+    # -fstrict-volatile-bitfields
+    # -mlongcalls
+    # -nostdlib
```
Comment (fedefrancescon) [inline CMakeLists.txt:34]:
```suggestion

```

**Pull request merged**